### ISSUE TITLE
PHPCD is full of errors (again)

### ISSUE THREAD

**Issue by shadowhand** [lvht/phpcd.vim#103]
Yesterday I ran `:PlugUpdate` and got a new version of phpcd. Now I am getting all the errors I used to get back in #53:

```
Error detected while processing function phpcd#UpdateIndex[6]..rpc#request:                                                                           
line    6:                                                                                                                                            
E906: not an open channel                                                                                                                             
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex[6]..rpc#request:                                                                           
line    6:                                                                                                                                            
E714: List required                                                                                                                                   
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex[6]..rpc#request:                                                                           
line    7:                                                                                                                                            
E121: Undefined variable: error                                                                                                                       
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex[6]..rpc#request:                                                                           
line    7:                                                                                                                                            
E15: Invalid expression: error                                                                                                                        
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex[6]..rpc#request:                                                                           
line   10:                                                                                                                                            
E121: Undefined variable: response                                                                                                                    
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex[6]..rpc#request:                                                                           
line   10:                                                                                                                                            
E15: Invalid expression: response                                                                                                                     
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex:                                                                                           
line    8:                                                                                                                                            
E121: Undefined variable: class                                                                                                                       
Press ENTER or type command to continue                                                                                                               
Error detected while processing function phpcd#UpdateIndex:                                                                                           
line    8:                                                                                                                                            
E15: Invalid expression: !nsuse.class 
...
```

Similar error messages continue for a very long time, or until I get sick of it and hit ESC.

Previous PHPCD version: e38a760
Current PHPCD version: bc0e5e8

I am reverting to e38a760 for the time being.

**Comment by xiaohuilam** [comment]
check directory into (just cd to) your vim/nvim's config plugged dir,
cd phpcd.vim,
try run composer update.

hope it works.

**Comment by shadowhand** [comment]
I've done that. The only change was installing php-parser.

**Comment by lvht** [comment]
@shadowhand please upgrade to the latest master commit and offer the ~/.phpcd.log info. Thans you.

**Comment by shadowhand** [comment]
% git checkout master
Previous HEAD position was e38a760... output more debug error message
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
% git pull
Current branch master is up to date.
% composer update
Loading composer repositories with package information
Updating dependencies (including require-dev)
Package operations: 1 install, 0 updates, 0 removals
  - Installing nikic/php-parser (v3.1.1): Loading from cache
Writing lock file
Generating optimized autoload files
```

Opened a file, used `SHIFT+K` to attempt a completion:

```
Error detected while processing function phpcd#CompletePHP[61]..phpcd#GetClassName[43]..phpcd#GetCallChainReturnType:                                                             
line   44:                                                                                                                                                                        
E701: Invalid type for len() 
```

Nothing in the log, except very old messages:

```
% tail -n15 ~/.phpcd.log
[2017-08-16 16:15:13] PHPCD.ERROR: Uninitialized string offset: 0 [{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/composer/ClassLoader.php","line":375,"function":"{closure}","args":[8,"Uninitialized string offset: 0","/Users/wgilk/.vim/plugged/phpcd.vim/vendor/composer/ClassLoader.php",375,{"class":"","ext":".php","logicalPathPsr4":".php"}]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/composer/ClassLoader.php","line":351,"function":"findFileWithExtension","class":"Composer\\Autoload\\ClassLoader","type":"->","args":["",".php"]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/composer/ClassLoader.php","line":321,"function":"findFile","class":"Composer\\Autoload\\ClassLoader","type":"->","args":[""]},{"function":"loadClass","class":"Composer\\Autoload\\ClassLoader","type":"->","args":[""]},{"function":"spl_autoload_call","args":[""]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/php/PHPID.php","line":195,"function":"__construct","class":"ReflectionClass","type":"->","args":["\\"]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/php/PHPID.php","line":40,"function":"getClassInfo","class":"PHPCD\\PHPID","type":"->","args":["\\"]},{"function":"update","class":"PHPCD\\PHPID","type":"->","args":["\\"]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/lvht/msgpack-rpc/src/AbstractServer.php","line":88,"function":"call_user_func_array","args":[["[object] (PHPCD\\PHPID: {})","update"],["\\"]]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/lvht/msgpack-rpc/src/AbstractServer.php","line":81,"function":"doRequest","class":"Lvht\\MsgpackRpc\\ForkServer","type":"->","args":[["[object] (PHPCD\\PHPID: {})","update"],["\\"]]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/lvht/msgpack-rpc/src/AbstractServer.php","line":47,"function":"onNotification","class":"Lvht\\MsgpackRpc\\ForkServer","type":"->","args":[[2,"update",["\\"]]]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/vendor/lvht/msgpack-rpc/src/ForkServer.php","line":32,"function":"onMessage","class":"Lvht\\MsgpackRpc\\ForkServer","type":"->","args":[[2,"update",["\\"]]]},{"file":"/Users/wgilk/.vim/plugged/phpcd.vim/php/main.php","line":39,"function":"loop","class":"Lvht\\MsgpackRpc\\ForkServer","type":"->","args":[]}] []
```

**Comment by lvht** [comment]
> Opened a file, used SHIFT+K to attempt a completion:

@shadowhand please offer the php code snippet. I will test the issue. Thanks.

**Comment by shadowhand** [comment]
@lvht https://gist.github.com/shadowhand/a4d77e4ae417b85b9074019bbf0d294b

**Comment by lvht** [comment]
I try it but not find any issue https://asciinema.org/a/JuUgAPbUTu7iwPuzCPJS4WEGd

**Comment by shadowhand** [comment]
Perhaps it is only a problem with vim and not with nvim?

**Comment by shadowhand** [comment]
On a related note... what options need to be set to have the preview window open and active for completions? I tried a number of things but the preview window never appears for me, even for completions that are working.

**Comment by lvht** [comment]
preview window will be opened automatically. no option is needed.

**Comment by xiaohuilam** [comment]
@duraki 

can you provide the debug trace file ~/.phpcd.log for @lvht to analysis what's happening?

**Comment by duraki** [comment]
Hey @xiaohuilam,

thanks, but unfortunately these logs were empty. I fixed it now; I think it was up to manual installation of `mgspack`. Used `pecl` instead of compiling and it didn't throw any error.

What also might help:
* Did a composer update in the `phpcd` bundle
* Did a `:PluginUpdate` in Vundle/Nvim

[Proof](https://i.imgur.com/IPNkKuY.png) :)

**Comment by zYeee** [comment]
@lvht 
I also got a lot of error when I typing ctrl+]

Error detected while processing function phpcd#JumpToDefinition[5]..phpcd#GetCurrentSymbolWithContext[43]..phpcd#GetCurrentNameSpace[1]..rpc#request:
line    6:
E906: not an open channel
E714: List required
line    7:
E121: Undefined variable: error
E15: Invalid expression: error
line   10:
E121: Undefined variable: response
E15: Invalid expression: response
Error detected while processing function phpcd#JumpToDefinition[5]..phpcd#GetCurrentSymbolWithContext[43]..phpcd#GetCurrentNameSpace:
line    3:
E121: Undefined variable: namespace
E15: Invalid expression: [nsuse.namespace, nsuse.imports]
Error detected while processing function phpcd#JumpToDefinition[5]..phpcd#GetCurrentSymbolWithContext:
line   43:
E714: List required
line   45:
E121: Undefined variable: current_namespace
E116: Invalid arguments for function phpcd#ExpandClassName(word, current_namespace, current_imports)
E15: Invalid expression: phpcd#ExpandClassName(word, current_namespace, current_imports)
line   50:
E121: Undefined variable: symbol
E15: Invalid expression: [symbol, context, symbol_namespace, current_imports]
Error detected while processing function phpcd#JumpToDefinition:
line    5:
E714: List required
line    6:
E121: Undefined variable: symbol
E15: Invalid expression: symbol == ''
line   10:
E121: Undefined variable: symbol
E116: Invalid arguments for function phpcd#LocateSymbol(symbol, symbol_context, symbol_namespace, current_imports)
E15: Invalid expression: phpcd#LocateSymbol(symbol, symbol_context, symbol_namespace, current_imports)
line   11:
E121: Undefined variable: symbol_file
E15: Invalid expression: symbol_file == '' || symbol_file == v:false
line   30:
E121: Undefined variable: symbol_line
E116: Invalid arguments for function str2nr(symbol_line) > 0
E15: Invalid expression: str2nr(symbol_line) > 0

Can you check about this? Thanks a lot :)

**Comment by lvht** [comment]
@zYeee these means your phpcd daemon does not bootstrap. you need to read the README and ensure all the system requirements has been installed.

**Comment by jmramos02** [comment]
Did you install it manually? If so, have you use composer install on the installation directory?

**Comment by jvmanji** [comment]
If someone have still similar problems, I've got such errors like described here today and I've figured out the cause simply by logging to file all arguments passed to _php/main.php_, and running this file "by hand". It occurred that "psr/log" was installed improperly due to some problems with composer cache. After doing:

```
rm -rf vendor/*
composer clear-cache
composer install
```

Everything is fine now.

Btw. I am using vim 8 (not neovim).
